src/Features/Core/Portable/PopulateSwitch/AbstractPopulateSwitchCodeFixProvider.cs

```diff
@@ -0,0 +1,122 @@
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeFixes.PopulateSwitch
+{
+    internal abstract class AbstractPopulateSwitchCodeFixProvider<TSwitchBlockSyntax, TExpressionSyntax, TSwitchSectionSyntax> : CodeFixProvider 
+            where TSwitchBlockSyntax : SyntaxNode
+            where TSwitchSectionSyntax : SyntaxNode
+            where TExpressionSyntax : SyntaxNode
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(IDEDiagnosticIds.PopulateSwitchDiagnosticId);
+
+        public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            context.RegisterCodeFix(
+                new MyCodeAction(
+                    FeaturesResources.AddMissingSwitchCases,
+                    c => AddMissingSwitchLabelsAsync(context)),
+                context.Diagnostics);
+
+            return Task.FromResult(true);
+        }
+
+        protected abstract TExpressionSyntax GetSwitchExpression(TSwitchBlockSyntax switchBlock);
+
+        protected abstract int InsertPosition(List<TSwitchSectionSyntax> sections);
+
+        protected abstract List<TSwitchSectionSyntax> GetSwitchSections(TSwitchBlockSyntax switchBlock);
+
+        protected abstract SyntaxNode NewSwitchNode(TSwitchBlockSyntax switchBlock, List<TSwitchSectionSyntax> sections);
+
+        protected abstract TSwitchBlockSyntax GetSwitchStatementNode(SyntaxNode root, TextSpan span);
+
+        protected abstract List<TExpressionSyntax> GetCaseLabels(TSwitchBlockSyntax switchBlock, out bool containsDefaultLabel);
+
+        private async Task<Document> AddMissingSwitchLabelsAsync(CodeFixContext context)
+        {
+            var document = context.Document;
+            var span = context.Span;
+            var cancellationToken = context.CancellationToken;
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var switchNode = GetSwitchStatementNode(root, span);
+
+            var enumType = (INamedTypeSymbol)model.GetTypeInfo(GetSwitchExpression(switchNode)).Type;
+
+            bool containsDefaultLabel;
+            var missingLabels = GetMissingLabels(switchNode, model, enumType, out containsDefaultLabel);
+
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            var switchExitStatement = generator.ExitSwitchStatement();
+            var statements = new List<SyntaxNode> { switchExitStatement };
+
+            var newSections = GetSwitchSections(switchNode);
```

Comment:
Note that this could just stay a ```SyntaxList<TSwitchSectionSyntax>```.  SyntaxList has methods like Add and Insert on it (though note that they do not mutate hte node in place, but instead return a new list).  In general i prefer just operating with syntax and not having to move between it's concepts and other collection types.